### PR TITLE
add mqtt to e2e-pubsub-test and fix wildcard subscription

### PR DESCRIPTION
**Pull request by thielepaul**
Signed-off-by: Paul Thiele <thielepaul@gmail.com>

# Description

Added mosquitto as a mqtt message broker to the e2e tests and added a mqtt subscription to e2e pubsub test.

## Issue reference

Please reference the issue this PR will close: #3920

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [x] Code compiles correctly
* [x] Created/updated tests
* [x] Unit tests passing
* [x] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

### CODE REVIEW

**Comment by thielepaul** [comment]
@artursouza can you trigger the e2e tests? The pubsub test should fail, and then pass after adding this: https://github.com/thielepaul/dapr/pull/1

**Comment by thielepaul** [comment]
@maintainers-dapr please approve the test workflow :)

**Review by yaron2 — APPROVED** [review]
lgtm

**Comment by artursouza** [comment]
I am reverting this change because it broke the test on Windows and the test should not be specific to a pubsub component.

**Comment by yaron2** [comment]
> I am reverting this change because it broke the test on Windows and the test should not be specific to a pubsub component.

Can we fix instead of revert?

File: tests/apps/pubsub-publisher/app.go [hunk 1]
```diff
@@ -85,12 +86,16 @@ func performPublish(w http.ResponseWriter, r *http.Request) {
...
+	if len(commandBody.PubSubName) == 0 {
+		commandBody.PubSubName = pubsubName
+	}
 
 	log.Printf("    commandBody.ContentType=%s", commandBody.ContentType)
 	log.Printf("    commandBody.Topic=%s", commandBody.Topic)
 	log.Printf("    commandBody.Data=%v", commandBody.Data)
 	log.Printf("    commandBody.Protocol=%s", commandBody.Protocol)
 	log.Printf("    commandBody.Metadata=%s", commandBody.Metadata)
+	log.Printf("    commandBody.PubSubName=%s", commandBody.PubSubName)
```
Comment (artursouza) [inline tests/apps/pubsub-publisher/app.go:98]: Please, do not make this test aware of pubsub. We keep the test generic.

File: tests/config/dapr_mqtt_pubsub.yaml [hunk 2]
```diff
@@ -0,0 +1,18 @@
+apiVersion: dapr.io/v1alpha1
+kind: Component
+metadata:
+  name: mqtt-pubsub
+spec:
+  type: pubsub.mqtt
+  version: v1
+  metadata:
+  - name: url
+    value: "tcp://mosquitto.dapr-tests.svc.cluster.local:1883"
```
Comment (artursouza) [inline tests/config/dapr_mqtt_pubsub.yaml:10]: Why bringing MQTT to the test?

File: tests/dapr_tests.mk [hunk 3]
```diff
@@ -313,6 +313,8 @@ setup-test-components: setup-app-configurations
 	$(KUBECTL) apply -f ./tests/config/app_topic_subscription_routing.yaml --namespace $(DAPR_TEST_NAMESPACE)
 	$(KUBECTL) apply -f ./tests/config/app_topic_subscription_routing_grpc.yaml --namespace $(DAPR_TEST_NAMESPACE)
 	$(KUBECTL) apply -f ./tests/config/app_pubsub_routing.yaml --namespace $(DAPR_TEST_NAMESPACE)
+	$(KUBECTL) apply -f ./tests/config/mosquitto.yaml --namespace $(DAPR_TEST_NAMESPACE)
```
Comment (artursouza) [inline tests/dapr_tests.mk:316]: Please, keep the test independent of mqtt, see how we have DAPR_TEST_PUBSUB as an env var to configure which pubsub to use for E2E tests. The e2e tests should pass independent of a component. For component specific test, we have certification tests in components-contrib.

**Comment by artursouza** [comment]
> > I am reverting this change because it broke the test on Windows and the test should not be specific to a pubsub component.
> 
> Can we fix instead of revert?

No, it is a rewrite.

**Comment by artursouza** [comment]
I know we don't do this for every PR, but in case of E2E test related PRs, we must run ok-to-test to make sure it passes on Windows too.

File: tests/config/dapr_mqtt_pubsub.yaml (hunk 2 above)
Comment (thielepaul) [reply tests/config/dapr_mqtt_pubsub.yaml:10]: while redis does support wildcards when using `psubscribe`, dapr uses `XGroupCreateMkStream` (https://github.com/dapr/components-contrib/blob/6e4139c7b1da21ea4a5d14d2e323a7ff37ee2d9f/pubsub/redis/redis.go#L172), which does not support wildcards.

File: tests/dapr_tests.mk (hunk 3 above)
Comment (thielepaul) [reply tests/dapr_tests.mk:316]: I think it is not really mqtt specific, as there are other brokers supported by dapr, which also support wildcards in the topics, but redis does not, so I am not sure what would be the best solution here.
If we move the test to components-contrib, this has the disadvantage, that we won't notice it if a PR to the main dapr repo breaks this again.

File: tests/apps/pubsub-publisher/app.go (hunk 1 above)
Comment (thielepaul) [reply tests/apps/pubsub-publisher/app.go:98]: yes, I agree why this is preferable. Maybe we can run the test for all supported brokers and make it somehow configurable, as not all brokers support the same features.

**Pull request merged**